crates/ibc-elc-client-cw/src/entrypoint.rs

```diff
@@ -0,0 +1,156 @@
+use crate::context::{Context, ExecutionContext};
+use crate::error::ContractError;
+use crate::msg::*;
+use crate::response::*;
+use cosmwasm_std::{to_json_binary, Binary, Deps, DepsMut, Env, MessageInfo, Response};
+use light_client::types::Any;
+use light_client::{ClientReader, HostContext, LightClient, UpdateClientResult};
+use prost::Message;
+
+pub trait Entrypoint {
+    type LightClient: LightClient;
+
+    fn get_status_from_client_state(any_client_state: Any) -> Result<String, ContractError>;
+    fn get_timestamp_from_consensus_state(any_consensus_state: Any) -> Result<u64, ContractError>;
+
+    fn instantiate(
+        lc: &Self::LightClient,
+        deps: DepsMut<'_>,
+        env: Env,
+        _info: MessageInfo,
+        msg: InstantiateMsg,
+    ) -> Result<Response, ContractError> {
+        let mut ctx = Context::new_mut(deps, env);
+        ctx.set_checksum(msg.checksum);
+
+        let any_client_state = Any::decode(&mut msg.client_state.as_slice())?;
+        let any_consensus_state = Any::decode(&mut msg.consensus_state.as_slice())?;
+
+        let res = lc.create_client(&ctx, any_client_state.clone(), any_consensus_state.clone())?;
+
+        ctx.store_client_state(res.height, any_client_state)?;
+        ctx.store_consensus_state(res.height, any_consensus_state)?;
+        ctx.store_update_meta(res.height, ctx.host_timestamp().into(), ctx.host_height())?;
+
+        Ok(Response::default().set_data(to_json_binary(&ContractResult::success())?))
+    }
+
+    fn sudo(
+        lc: &Self::LightClient,
+        deps: DepsMut<'_>,
+        env: Env,
+        msg: SudoMsg,
+    ) -> Result<Response, ContractError> {
+        let mut ctx = Context::new_mut(deps, env);
+
+        let result = match msg {
+            SudoMsg::UpdateState(msg) => {
+                let any_message = Any::decode(msg.client_message.as_slice())?;
+                let res = match lc.update_client(&ctx, ctx.client_id().clone(), any_message)? {
+                    UpdateClientResult::UpdateState(d) => d,
+                    _ => panic!("unexpected non-UpdateState client message"),
+                };
+
+                ctx.store_client_state(res.height, res.new_any_client_state)?;
+                ctx.store_consensus_state(res.height, res.new_any_consensus_state)?;
+                ctx.store_update_meta(res.height, ctx.host_timestamp().into(), ctx.host_height())?;
+
+                ContractResult::success().heights(vec![res.height])
+            }
+            SudoMsg::UpdateStateOnMisbehaviour(msg) => {
+                let any_message = Any::decode(msg.client_message.as_slice())?;
+                let res = match lc.update_client(&ctx, ctx.client_id().clone(), any_message)? {
+                    UpdateClientResult::Misbehaviour(d) => d,
+                    _ => panic!("unexpected non-Misbehaviour client message"),
```

Comment:
If this panic can occur due to the tx submitted by the relayer, we should return an error instead.